### PR TITLE
GPU: Use lazy checks for specialization state

### PR DESCRIPTION
**Pull request by riperiperi**
This PR adds a new class, the SpecializationStateUpdater, that allows elements of specialization state to be updated individually, and signal the state is checked when it changes between draws, instead of building and checking it on every draw. This also avoids building spec state switching shaders.

Most state updates have been moved behind the shader state update, so that their specialization state updates make it in before shaders are fetched.

Downside: Fields in GpuChannelGraphicsState are no longer readonly. To counteract copies that might be caused this I pass it as `ref` when possible, though maybe `in` would be better? Not really sure about the quirks of `in` and the difference probably won't show on a benchmark.

The result is around 2 extra FPS on SMO here in the usual spot. Not much right now, but it will remove costs when we're doing more expensive specialization checks, such as fragment output type specialization for macos. It may also help more on other games with more draws.

### CODE REVIEW

File: Ryujinx.Graphics.Gpu/Engine/Threed/SpecializationStateUpdater.cs [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+                _graphics.TessellationMode = value;
+
+                Signal();
+            }
+        }
+
+        /// <summary>
+        /// Indicates whether alpha-to-coverage is enabled.
+        /// </summary>
+        /// <param name="value">The new value</param>
```
Comment (gdkchan) [inline Ryujinx.Graphics.Gpu/Engine/Threed/SpecializationStateUpdater.cs:121]: Wrong XML docs (parameters are `enable` and `ditherEnable`).

File: Ryujinx.Graphics.Gpu/Engine/Threed/StateUpdater.cs [hunk 2]
```diff
@@ -16,9 +17,9 @@ namespace Ryujinx.Graphics.Gpu.Engine.Threed
     /// </summary>
     class StateUpdater
     {
-        public const int ShaderStateIndex = 16;
+        public const int ShaderStateIndex = 26;
```
Comment (gdkchan) [inline Ryujinx.Graphics.Gpu/Engine/Threed/StateUpdater.cs:20]: I think we should expose properties with those values and query the correct value by name on the class constructor. Thee current approach is too error prone. Doesn't need to be done right now though.

**Review by gdkchan — COMMENTED** [review]
`GetGraphicsState` and `GetPoolState` on `StateUpdater` seems to be unused  now, so I guess they can be removed?

Also, it seems that the state on `SpecializationStateUpdater` is updated when each state update method is called. Since that happens after the the specialization checks that would invalidate the current shader, isn't it just comparing state from the previous draw?

**Comment by riperiperi** [comment]
I assumed since the spec check was done there before, the state update had already happened, but that wouldn't make sense as it sets the shader dirty flag. I've updated it to reload after (UpdateShaderState rather than set the flag), and added a comment about the behaviour should not change _vtgWritesRtLayer.

**Review by gdkchan — APPROVED** [review]
lgtm, but it's a bit unfortunate that there's now 2 places that might cause the shader to be updated (one for state spec and other for textures spec).

**Comment by riperiperi** [comment]
It kind of sucks, but I can at least explain how it works now:

- UpdateShaderState is up first. If the shader state is dirty, it does the full check with textures and everything.
- Check before bindings to see if graphics state is dirty after processing updates, and reloads the shader if it is different. If the shader was changed, this will always do nothing, as the state dirty flag was reset as it used the current state fo find the shader earlier.
  - I wasn't comfortable combining this one with the texture bindings check, as that would mean that none of the bindings are allowed to change due to spec state. I'm not sure that would always be true for graphics spec state now or in the future.
- Texture bindings check as they are fetched, and reload the shader if different. If the shader was changed by _either_ of the above two things, the texture state will already match.

At least it should only ever reload once, it's just a case of when.

**Review by marysaka — APPROVED** [review]

**Pull request merged**